zkevm-circuits/src/evm_circuit/util/constraint_builder.rs

```diff
@@ -1169,33 +1121,149 @@ impl<'a, F: FieldExt> ConstraintBuilder<'a, F> {
     }
 
     pub(crate) fn add_constraints(&mut self, constraints: Vec<(&'static str, Expression<F>)>) {
-        self.cb.add_constraints(constraints);
+        for (name, constraint) in constraints {
+            self.add_constraint(name, constraint);
+        }
     }
 
     pub(crate) fn add_constraint(&mut self, name: &'static str, constraint: Expression<F>) {
-        self.cb.add_constraint(name, constraint);
+        let constraint = match &self.condition {
+            Some(condition) => condition.clone() * constraint,
+            None => constraint,
+        };
+
+        let constraint = self.split_expression(name, constraint, MAX_DEGREE - IMPLICIT_DEGREE);
+
+        self.validate_degree(constraint.degree(), name);
+        self.constraints.push((name, constraint));
     }
 
     pub(crate) fn add_constraint_first_step(
         &mut self,
         name: &'static str,
         constraint: Expression<F>,
     ) {
-        let constraint = match &self.cb.condition {
+        let constraint = match &self.condition {
             Some(condition) => condition.clone() * constraint,
             None => constraint,
         };
-        // Add 1 more degree due to the selector
-        self.validate_degree(constraint.degree() + 1, name);
+        self.validate_degree(constraint.degree(), name);
         self.constraints_first_step.push((name, constraint));
     }
 
     pub(crate) fn add_lookup(&mut self, name: &'static str, lookup: Lookup<F>) {
-        let lookup = match &self.cb.condition {
+        let lookup = match &self.condition {
             Some(condition) => lookup.conditional(condition.clone()),
             None => lookup,
         };
-        self.validate_degree(lookup.degree() + LOOKUP_DEGREE, name);
-        self.lookups.push((name, lookup));
+
+        let compressed_expr = self.split_expression(
+            "Lookup compression",
+            rlc::expr(&lookup.input_exprs(), self.power_of_randomness),
+            MAX_DEGREE - IMPLICIT_DEGREE,
+        );
+        self.store_expression(
+            name,
+            compressed_expr,
+            CellType::Lookup(lookup.table()),
+            true,
+        );
+    }
+
+    pub(crate) fn store_expression(
+        &mut self,
+        name: &'static str,
+        expr: Expression<F>,
+        cell_type: CellType,
+        allow_reuse: bool,
```

Comment:
Yep not needed anymore, but yeah I thought it could still be useful somehow but, at least how things currently work, I can't find any use cases so removed it.

And great idea about the duplicated lookups, will do!